### PR TITLE
feat: add kubelet flag checks

### PR DESCRIPTION
**Pull request by frezbo**
Add kubelet flag checks.
Add k8s 1.27 upgrade checks.

This is a breaking change in how `upgrade.NewChecks` is created.

### CODE REVIEW

File: kubernetes/upgrade/checks.go [hunk 1]
```diff
@@ -172,10 +231,10 @@ func (checks *Checks) Run(ctx context.Context) error {
 }
 
 // PopulateRemovedCLIFlags populates the removed flags.
-func (e *ComponentRemovedItemsError) PopulateRemovedCLIFlags(node, component string, apiServerCLIFlags []string, removedFlags []string) {
+func (e *ComponentRemovedItemsError) PopulateRemovedCLIFlags(node, component string, cliFlags []string, removedFlags []string) {
 	for _, removedFlag := range removedFlags {
-		if slices.Contains(apiServerCLIFlags, func(s string) bool {
-			return strings.HasPrefix(s, "--"+removedFlag)
+		if slices.Contains(cliFlags, func(s string) bool {
+			return "--"+removedFlag == strings.Split(s, "=")[0]
```
Comment (frezbo) [inline kubernetes/upgrade/checks.go:237]: Fixes a bug in how cli flags were compared

File: kubernetes/upgrade/checks_test.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func TestK8sComponentRemovedItemsWithError(t *testing.T) {
 				"/usr/local/bin/kube-apiserver",
 				"--bind-address=0.0.0.0",
 				"--insecure-port=0",
-				"--feature-gates=RotateKubeletServerCertificate=true,CSIVolumeFSGroupPolicy",
+				"--feature-gates=RotateKubeletServerCertificate=true,CSIVolumeFSGroupPolicy=true",
```
Comment (frezbo) [inline kubernetes/upgrade/checks_test.go:71]: the second feature gate was missing the bool argument previosuly

File: kubernetes/upgrade/checks_test.go [hunk 3]
```diff
@@ -149,7 +149,161 @@ func TestK8sComponentRemovedItemsWithError(t *testing.T) {
...
+				},
+			},
+		}
+
+		require.NoError(t, resourceState.Create(ctx, cfg))
+	}
+
+	cfg := k8s.NewKubeletSpec(k8s.NamespaceName, k8s.KubeletID)
+	cfg.TypedSpec().Args = []string{
+		"--container-runtime=containerd",
```
Comment (frezbo) [inline kubernetes/upgrade/checks_test.go:293]: ``--container-runtime` and `--container-runtime-endpoint`makes sure we are checking the exact value for a flag key and not just a prefix

**Comment by frezbo** [comment]
```bash
_out/talosctl-linux-amd64 -n 10.5.0.2 upgrade-k8s --to=v1.27.0-beta.0
automatically detected the lowest Kubernetes version 1.26.2
discovered controlplane nodes ["10.5.0.2"]
discovered worker nodes ["10.5.0.3"]
checking for removed Kubernetes component flags
checking for removed Kubernetes API resource versions

NODE       COMPONENT   REMOVED FLAG
10.5.0.3   kubelet     master-service-namespace
```

fails the check when `master-service-namespace` is added to kubelet extraArgs

File: kubernetes/upgrade/checks.go [hunk 4]
```diff
@@ -118,11 +123,52 @@ func NewChecks(path *Path, state state.State, k8sConfig *rest.Config, controlPla
...
 			},
+			// https://kubernetes.io/blog/2023/03/17/upcoming-changes-in-kubernetes-v1-27/
+			"1.26->1.27": {
+				kubeAPIServerChecks: apiServerCheck{
+					removedAPIResources: []string{
+						"csistoragecapacities.v1beta1.storage.k8s.io",
+					},
+					componentCheck: componentCheck{
+						removedFlags: []string{
+							"master-service-namespace",
```
Comment (utkuozdemir) [inline kubernetes/upgrade/checks.go:134]: I checked this one, and it seems this flag was actually removed from kubelet, not kube-apiserver and [this page](https://kubernetes.io/blog/2023/03/17/upcoming-changes-in-kubernetes-v1-27/#removal-of-master-service-namespace-command-line-argument) has wrong information, unless if I am missing something.

Comment (frezbo) [reply kubernetes/upgrade/checks.go:134]: both are removed, apiserver and kubelet https://github.com/kubernetes/kubernetes/pull/116015 see this PR

Comment (frezbo) [reply kubernetes/upgrade/checks.go:134]: the page actually missed information on removed kubelet flags :sweat_smile:

Comment (frezbo) [reply kubernetes/upgrade/checks.go:134]: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#other-cleanup-or-flake it's in the changelog though :sweat_smile:

Comment (utkuozdemir) [reply kubernetes/upgrade/checks.go:134]: Ah, so it seems it was removed from `kube-apiserver` already on `1.25->1.26`, and now it is also removed from `kubelet` on `1.26->1.27`.

Comment (frezbo) [reply kubernetes/upgrade/checks.go:134]: oh it was removed way back, in k8s 1.7, i missed that, good catch, I'll fixup the PR, yes, the link is wrong it seems

Comment (utkuozdemir) [reply kubernetes/upgrade/checks.go:134]: cool. just a small note: I downloaded the binaries of apiserver of various versions and checked myself - the flag seems to be removed from it exactly on `1.25->1.26`, not way back in 1.7.

Comment (frezbo) [reply kubernetes/upgrade/checks.go:134]: sure in 1.26? https://github.com/kubernetes/kubernetes/pull/114446 this says 1.27 thinking

Comment (utkuozdemir) [reply kubernetes/upgrade/checks.go:134]: 🤔 

```
$ $ ./kube-apiserver-1.25 --version
Kubernetes v1.25.0
$ ./kube-apiserver-1.25 --help | grep master-service
      --master-service-namespace string                DEPRECATED: the namespace from which the Kubernetes master services should be injected into pods. (default "default")
$ ./kube-apiserver-1.26 --version
Kubernetes v1.26.3
$ ./kube-apiserver-1.26 --help | grep master-service
$ ./kube-apiserver-1.27 --version
Kubernetes v1.27.0-beta.0
$ ./kube-apiserver-1.27 --help | grep master-service
$
```

Comment (frezbo) [reply kubernetes/upgrade/checks.go:134]: yeh, seems it got removed in 1.26, but the PR mentions 1.26, seems a little confusion

File: kubernetes/upgrade/checks.go (hunk 1 above)
Comment (smira) [reply kubernetes/upgrade/checks.go:239]: it might be eaiser to use `strings.Cut`, it's more to the point here I think

**Review by smira — APPROVED** [review]

File: kubernetes/upgrade/checks_test.go [hunk 5]
```diff
@@ -47,7 +47,7 @@ func TestK8sComponentRemovedItemsNoError(t *testing.T) {
 	path, err := upgrade.NewPath("1.24.3", "1.25.0")
 	require.NoError(t, err)
 
-	checks, err := upgrade.NewChecks(path, resourceState, nil, []string{"10.5.0.2"}, t.Logf)
+	checks, err := upgrade.NewChecks(path, resourceState, nil, []string{"10.5.0.2"}, []string{}, t.Logf)
```
Comment (DmitriyMV) [inline kubernetes/upgrade/checks_test.go:50]:
```suggestion
	checks, err := upgrade.NewChecks(path, resourceState, nil, []string{"10.5.0.2"}, nil, t.Logf)
```

**Pull request merged**